contracts/contracts/strategies/sonic/SonicValidatorDelegator.sol

```diff
@@ -0,0 +1,396 @@
+
+        withdrawId = nextWithdrawId++;
+
+        withdrawals[withdrawId] = WithdrawRequest(
+            validatorId,
+            undelegateAmount,
+            block.timestamp
+        );
+        pendingWithdrawals += undelegateAmount;
+
+        sfc.undelegate(validatorId, withdrawId, undelegateAmount);
+
+        emit Undelegated(withdrawId, validatorId, undelegateAmount);
+    }
+
+    /**
+     * @notice Withdraw native S from a previously undelegated validator.
+     * The native S is wrapped wS and transferred to the Vault.
+     * @param withdrawId The withdraw ID returned from `undelegate` and emitted in `Undelegated`.
+     * @return withdrawnAmount The amount of Sonic (S) withdrawn.
+     * This can be less than the undelegated amount in the event of slashing.
+     */
+    function withdrawFromSFC(uint256 withdrawId)
+        external
+        onlyRegistratorOrStrategist
+        nonReentrant
+        returns (uint256 withdrawnAmount)
+    {
+        require(withdrawId < nextWithdrawId, "Invalid withdrawId");
+
+        // Can still withdraw even if the validator is no longer supported
+        // Load the withdrawal from storage into memory
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(!isWithdrawnFromSFC(withdrawId), "Already withdrawn");
+
+        withdrawals[withdrawId].undelegatedAmount = 0;
+        pendingWithdrawals -= withdrawal.undelegatedAmount;
+
+        uint256 sBalanceBefore = address(this).balance;
+
+        sfc.withdraw(withdrawal.validatorId, withdrawId);
+
+        // Save state to storage
+        withdrawnAmount = address(this).balance - sBalanceBefore;
+
+        // Wrap Sonic (S) to Wrapped Sonic (wS)
+        IWrappedSonic(wrappedSonic).deposit{ value: withdrawnAmount }();
+
+        // Transfer the Wrapped Sonic (wS) to the Vault
+        _withdraw(vaultAddress, wrappedSonic, withdrawnAmount);
+
+        // withdrawal.undelegatedAmount & withdrawnAmount can differ in case of slashing
+        emit Withdrawn(
+            withdrawId,
+            withdrawal.validatorId,
+            withdrawal.undelegatedAmount,
+            withdrawnAmount
+        );
+    }
+
+    /// @notice returns a bool whether a withdrawalId has already been withdrawn or not
+    function isWithdrawnFromSFC(uint256 withdrawId) public view returns (bool) {
+        WithdrawRequest memory withdrawal = withdrawals[withdrawId];
+        require(withdrawal.validatorId > 0, "Invalid withdrawId");
+        return withdrawal.undelegatedAmount == 0;
+    }
+
+    /**
+     * @notice Restake any pending validator rewards for all supported validators
+     * @param validatorIds List of Sonic validator IDs to restake rewards
+     */
+    function restakeRewards(uint256[] calldata validatorIds)
+        external
+        nonReentrant
+    {
+        for (uint256 i = 0; i < validatorIds.length; ++i) {
+            require(
```

Comment:
❓Do we need this require? It might be handy to be able to collect rewards from validators we've already closed out. In this function we are collecting money only, under non-public control.

Reply:
https://github.com/OriginProtocol/origin-dollar/pull/2335/commits/c76f2b92167b63172a934a39efab38d7c9d9e426